### PR TITLE
refactor: update enhanced meshing interface

### PR DESCRIPTION
**Pull request by prmukherj**
1. Add documentation and testing for 2D Meshing.
2. Refactoring of entire Meshing workflow.

### CODE REVIEW

File: src/ansys/fluent/core/workflow.py [hunk 1]
```diff
@@ -984,17 +984,22 @@ def _add_child(self, state: Optional[dict] = None) -> None:
...
+            Flag to defer update.
         """
-        self._add_child(state)
-        self._task.AddChildAndUpdate()
-        return self.last_child()
+        if defer_update is None:
+            self._add_child(state)
+            self._task.AddChildAndUpdate()
+            return self.last_child()
+        else:
```
Comment (prmukherj) [inline src/ansys/fluent/core/workflow.py:1001]: state might not be None here.

File: src/ansys/fluent/core/workflow.py [hunk 2]
```diff
@@ -979,21 +979,26 @@ def _add_child(self, state: Optional[dict] = None) -> None:
...
             Optional state.
+        defer_update: Optional[bool]
+            Flag to defer update.
         """
         self._add_child(state)
-        self._task.AddChildAndUpdate()
+        if defer_update is None:
+            self._task.AddChildAndUpdate()
+        else:
+            return self._task.AddChildAndUpdate(DeferUpdate=defer_update)
```
Comment (mkundu1) [inline src/ansys/fluent/core/workflow.py:1001]: Are the behaviours `AddChildAndUpdate()` and `AddChildAndUpdate(DeferUpdate=False)` different? If not, the default value of `defer_update` can be set to False.

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst [hunk 3]
```diff
@@ -572,6 +572,220 @@ Switch to solution mode
 
     solver = meshing.switch_to_solver()
 
+
+2D meshing workflow
+-------------------
+Use the **2D Meshing** workflow to mesh specific two dimensional geometries.
```
Comment (seanpearsonuk) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:578]: Just wondering why this is bold and Meshing has capital M in these two files.

File: src/ansys/fluent/core/workflow.py [hunk 4]
```diff
@@ -979,21 +979,23 @@ def _add_child(self, state: Optional[dict] = None) -> None:
...
 
-    def add_child_and_update(self, state=None):
+    def add_child_and_update(self, state=None, defer_update=False):
         """Add a child to this CompoundTask and update.
 
         Parameters
         ----------
         state : Optional[dict]
             Optional state.
+        defer_update: Optional[bool]
```
Comment (mkundu1) [inline src/ansys/fluent/core/workflow.py:994]:
```suggestion
        defer_update: bool, default False
```

**Review by mkundu1 — APPROVED** [review]

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst [hunk 5]
```diff
@@ -572,6 +572,220 @@ Switch to solution mode
 
     solver = meshing.switch_to_solver()
 
+
+2D meshing workflow
+-------------------
+Use the **2D** meshing workflow to mesh specific two dimensional geometries.
+The following example shows you how to use the 2D Meshing workflow.
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:579]:
```suggestion
Use the **2D** meshing workflow to mesh specific two-dimensional geometries.
The following example shows how to use the 2D Meshing workflow.
```
I don't think we need to bold 2D as the title already ponts out the topic is about 2D.

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst [hunk 6]
```diff
@@ -572,6 +572,220 @@ Switch to solution mode
...
+            r"CurvatureNormalAngle": 20,
+            r"MaxSize": 2000,
+            r"MinSize": 5,
+            r"SizeFunctions": r"Curvature",
+        }
+    )
+    meshing.workflow.TaskObject["Define Global Sizing"].Execute()
+
+Adding BOI
+~~~~~~~~~~
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:634]:
```suggestion
Add body of influence
~~~~~~~~~~~~~~
```
I don't think we can assume someone will know what BOI means, especially in a title. The code itself implicitly defines it. However, a glance at the heading would likely be confusing.

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst [hunk 7]
```diff
@@ -572,6 +572,220 @@ Switch to solution mode
...
+            r"NumberOfLayers": 4,
+            r"OffsetMethodType": r"aspect-ratio",
+        }
+    )
+    meshing.workflow.TaskObject["Add 2D Boundary Layers"].AddChildAndUpdate(
+        DeferUpdate=False
+    )
+
+Generate the surface mesh
+~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:708]:
```suggestion
Generate surface mesh
~~~~~~~~~~~~~~~~~~~~~
```
Removing the article "the" as articles aren't used in the other headings.

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst [hunk 8]
```diff
@@ -572,6 +572,220 @@ Switch to solution mode
...
+    meshing.workflow.TaskObject["Load CAD Geometry"].Arguments.set_state(
+        {
+            r"FileName": import_file_name,
+            r"LengthUnit": r"mm",
+            r"Refaceting": {
+                r"Refacet": False,
+            },
+        }
+    )
+    meshing.workflow.TaskObject["Load CAD Geometry"].Execute()
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:603]:
```suggestion
    meshing.workflow.TaskObject["Load CAD geometry."].Execute()
```
I'm assuming the message string can be edited to use sentence case and concluding punctuation.

Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:603]: For all these descriptions in the quotation marks, can they be edited to use sentence case and concluding punctuation? I'm not sure if they are message strings shown to the user. If so, they should be written as a sentence that concludes with a period.

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst [hunk 9]
```diff
@@ -572,6 +572,220 @@ Switch to solution mode
...
+
+Switch to solution mode
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Switching to solver is not allowed in 2D Meshing mode.
+
+
 Sample use of CommandArguments
 ------------------------------
 This simple example shows you how to use the ``CommandArgument`` attributes and explicit
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:791]:
```suggestion
This simple example shows how to use the ``CommandArgument`` attributes and explicit
```

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst [hunk 10]
```diff
@@ -572,6 +572,220 @@ Switch to solution mode
...
+    meshing.workflow.TaskObject["Export Fluent 2D Mesh"].Execute()
+
+Switch to solution mode
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Switching to solver is not allowed in 2D Meshing mode.
+
+
 Sample use of CommandArguments
 ------------------------------
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:790]:
```suggestion
Sample use of ``CommandArguments`` arguments
--------------------------------------------
```

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst [hunk 11]
```diff
@@ -572,6 +572,220 @@ Switch to solution mode
...
+        {
+            r"FileName": r"mesh1.msh.h5",
+        }
+    )
+    meshing.workflow.TaskObject["Export Fluent 2D Mesh"].Execute()
+
+Switch to solution mode
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Switching to solver is not allowed in 2D Meshing mode.
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:786]:
```suggestion
Switching to solver mode is not allowed in 2D Meshing mode.
```
The title says "Switch to solution mode," yet the sentence says solver mode? Mistake maybe?

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst (hunk 10 above)
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:790]:
```suggestion
Sample use of ``CommandArguments`` attributes
---------------------------------------------
```

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst (hunk 9 above)
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/meshing_workflows.rst:791]:
```suggestion
This simple example shows how to use the ``CommandArgument`` attributes and explicit
```
Other writers have asked to drop that we drop the unnecessary "you".

File: doc/source/user_guide/meshing_workflow/new_meshing_workflows.rst [hunk 12]
```diff
@@ -453,3 +453,163 @@ Switch to solution mode
 .. code:: python
 
     solver = meshing.switch_to_solver()
+
+
+2D meshing workflow
+-------------------
+Use the **2D** meshing workflow to mesh specific two dimensional geometries.
+The following example shows you how to use the 2D Meshing workflow.
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/new_meshing_workflows.rst:461]:
```suggestion
The following example shows you how to use the 2D meshing workflow.
```

File: doc/source/user_guide/meshing_workflow/new_meshing_workflows.rst [hunk 13]
```diff
@@ -453,3 +453,163 @@ Switch to solution mode
 .. code:: python
 
     solver = meshing.switch_to_solver()
+
+
+2D meshing workflow
+-------------------
+Use the **2D** meshing workflow to mesh specific two dimensional geometries.
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/new_meshing_workflows.rst:460]:
```suggestion
Use the **2D** meshing workflow to mesh specific two-dimensional geometries.
```
Again, I'd advise not bolding 2D. Why does this topic seem repetitive?!

File: doc/source/user_guide/meshing_workflow/new_meshing_workflows.rst [hunk 14]
```diff
@@ -453,3 +453,163 @@ Switch to solution mode
...
+.. code:: python
+
+    two_dim_mesh.add_2d_boundary_layers.add_child = "yes"
+    two_dim_mesh.add_2d_boundary_layers.bl_control_name = "aspect-ratio_1"
+    two_dim_mesh.add_2d_boundary_layers.number_of_layers = 4
+    two_dim_mesh.add_2d_boundary_layers.offset_method_type = "aspect-ratio"
+    two_dim_mesh.add_2d_boundary_layers.add_child_and_update(defer_update=False)
+
+Generate the surface mesh
+~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/new_meshing_workflows.rst:559]:
```suggestion
Generate surface mesh
~~~~~~~~~~~~~~~~~~~~~
```

File: doc/source/user_guide/meshing_workflow/new_meshing_workflows.rst [hunk 15]
```diff
@@ -453,3 +453,163 @@ Switch to solution mode
...
+    )
+    two_dim_mesh.generate_initial_surface_mesh.surface2_d_preferences.merge_face_zones_based_on_labels = (
+        "no"
+    )
+    two_dim_mesh.generate_initial_surface_mesh.surface2_d_preferences.show_advanced_options = (
+        True
+    )
+    two_dim_mesh.generate_initial_surface_mesh()
+
+Switch to solution mode
```
Comment (PipKat) [inline doc/source/user_guide/meshing_workflow/new_meshing_workflows.rst:612]: Same question here as I find the title versus the sentence confusing.

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 16]
```diff
@@ -25,8 +28,12 @@ def __init__(
             Underlying workflow object.
         meshing : PyMenuGeneric
             Meshing object.
+        fluent_version: FluentVersion
+            Version information of the current fluent session.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:32]:
```suggestion
            Version of Fluent to use in this session.
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 17]
```diff
@@ -46,14 +56,45 @@ def __init__(
             Underlying workflow object.
         meshing : PyMenuGeneric
             The meshing object.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:58]:
```suggestion
            Meshing object.
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 18]
```diff
@@ -46,14 +56,45 @@ def __init__(
...
+        name: str
+            Workflow name to initialize it.
+        identifier: str
+            Workflow name to identify it from global settings.
+        fluent_version: FluentVersion
+            Version information of the current fluent session.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:64]:
```suggestion
            Workflow name for this instance.
        identifier: str
            Workflow name to identify it from global settings.
        fluent_version: FluentVersion
            Version of Fluent to use in this session.
```
I'm a bit confused by the description for the "Identifier" parameter. Maybe "Name of the workflow in the global settings." I'm probably misunderstanding the purpose, but you can see how clarifying the description would be helpful to users.

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 19]
```diff
@@ -62,24 +103,16 @@ def __init__(self, workflow: PyMenuGeneric, meshing: PyMenuGeneric) -> None:
             The underlying workflow object.
         meshing : PyMenuGeneric
             The meshing object.
+        fluent_version: FluentVersion
+            Version information of the current fluent session.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:107]:
```suggestion
            Underlying workflow object.
        meshing : PyMenuGeneric
            Meshing object.
        fluent_version: FluentVersion
            Version of Fluent to use in this session.
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 20]
```diff
@@ -104,27 +138,19 @@ def __init__(
             The part-management object.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:138]:
```suggestion
            Part management object.
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 21]
```diff
@@ -104,27 +138,19 @@ def __init__(
...
         pm_file_management : PyMenuGeneric
             The part-management file-management object.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:140]:
```suggestion
            File management object in the part management file.
```
Leading article "The" should be removed from the all previous brief parameter description.

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 22]
```diff
@@ -104,27 +138,19 @@ def __init__(
...
+        fluent_version: FluentVersion
+            Version information of the current fluent session.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:142]:
```suggestion
            Version of Fluent to use in this session.
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 23]
```diff
@@ -160,30 +191,27 @@ def __init__(self, workflow: PyMenuGeneric, meshing: PyMenuGeneric) -> None:
             The underlying workflow object.
         meshing : PyMenuGeneric
             The meshing object.
+        fluent_version: FluentVersion
+            Version information of the current fluent session.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:195]:
```suggestion
            Version of Fluent to use in this session.
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 24]
```diff
@@ -160,30 +191,27 @@ def __init__(self, workflow: PyMenuGeneric, meshing: PyMenuGeneric) -> None:
             The underlying workflow object.
         meshing : PyMenuGeneric
             The meshing object.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:193]:
```suggestion
            Underlying workflow object.
        meshing : PyMenuGeneric
            Meshing object.
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 25]
```diff
@@ -192,21 +220,23 @@ def __init__(self, workflow: PyMenuGeneric, meshing: PyMenuGeneric) -> None:
...
+        fluent_version: FluentVersion
+            Version information of the current fluent session.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:224]:
```suggestion
        fluent_version: FluentVersion
            Version of Fluent to use in this session..
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 26]
```diff
@@ -192,21 +220,23 @@ def __init__(self, workflow: PyMenuGeneric, meshing: PyMenuGeneric) -> None:
             The underlying workflow object.
         meshing : PyMenuGeneric
             The meshing object.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:222]:
```suggestion
            Underlying workflow object.
        meshing : PyMenuGeneric
            Meshing object.
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 27]
```diff
@@ -192,21 +220,23 @@ def __init__(self, workflow: PyMenuGeneric, meshing: PyMenuGeneric) -> None:
...
+            workflow=workflow,
+            meshing=meshing,
+            name="Topology Based Meshing",
+            identifier="EnablePrimeMeshing",
+            fluent_version=fluent_version,
+        )
+
+
+class WorkflowMode(Enum):
+    """Enumerates over supported Fluent meshing workflow modes."""
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:236]:
```suggestion
    """Provides an enum of supported Fluent meshing workflow modes."""
```

File: src/ansys/fluent/core/session_base_meshing.py [hunk 28]
```diff
@@ -152,51 +146,56 @@ def _workflow_se(self):
...
                 self.meshing,
                 self.PartManagement,
                 self.PMFileManagement,
+                self.get_fluent_version(),
             )
         return self._ft_workflow
 
     @property
     def two_dimensional_meshing_workflow(self):
         """Datamodel root of workflow exposed in object-oriented manner."""
```
Comment (PipKat) [inline src/ansys/fluent/core/session_base_meshing.py:182]:
```suggestion
        """Data model root of the workflow exposed in an object-oriented manner."""
```

File: src/ansys/fluent/core/workflow.py [hunk 29]
```diff
@@ -979,21 +981,33 @@ def _add_child(self, state: Optional[dict] = None) -> None:
...
-            self.arguments.set_state(state)
+            self.arguments.update_dict(state)
         else:
             state.update({"AddChild": "yes"})
-            self._task.Arguments.set_state(state)
+            self._task.Arguments.update_dict(state)
 
-    def add_child_and_update(self, state=None):
+    def add_child_and_update(self, state=None, defer_update=None):
         """Add a child to this CompoundTask and update.
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:990]:
```suggestion
        """Add a child to the compound task and update.
```

File: src/ansys/fluent/core/workflow.py [hunk 30]
```diff
@@ -979,21 +981,33 @@ def _add_child(self, state: Optional[dict] = None) -> None:
...
 
         Parameters
         ----------
         state : Optional[dict]
             Optional state.
+        defer_update: bool, default False
+            Flag to defer update.
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:997]:
```suggestion
        defer_update : bool, default: False
            Whether to defer the update.
```

File: src/ansys/fluent/core/workflow.py [hunk 31]
```diff
@@ -979,21 +981,33 @@ def _add_child(self, state: Optional[dict] = None) -> None:
...
         self._add_child(state)
-        self._task.AddChildAndUpdate()
+        if self._fluent_version >= FluentVersion.v241:
+            if defer_update is None:
+                defer_update = False
+            self._task.AddChildAndUpdate(DeferUpdate=defer_update)
+        else:
+            if defer_update is not None:
+                warnings.warn(
+                    "'defer_update' is supported for Fluent versions 24.1 onwards.",
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:1007]:
```suggestion
                    " The 'defer_update()' method is supported in Fluent 2024 R1 and later.",
```

**Review by PipKat — APPROVED** [review]
Suggested edits to ensure content follows our style guide lines.

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst (hunk 5 above)
Comment (prmukherj) [reply doc/source/user_guide/meshing_workflow/meshing_workflows.rst:579]: @PipKat, the bold was done to be consistent with other workflow types.

File: doc/source/user_guide/meshing_workflow/meshing_workflows.rst (hunk 8 above)
Comment (prmukherj) [reply doc/source/user_guide/meshing_workflow/meshing_workflows.rst:603]: These are basically display names of tasks by Fluent. These should be as it is. Thank you :)

**Review by hpohekar — APPROVED** [review]

**Pull request merged**